src/libaktualizr/http/httpclient.cc

```diff
@@ -352,4 +352,43 @@ curl_slist* HttpClient::curl_slist_dup(curl_slist* sl) {
   return new_list;
 }
 
+/* Locking for curl share instance */
+static void curl_share_lock_cb(CURL* handle, curl_lock_data data, curl_lock_access access, void* userptr) {
+  (void)handle;
+  (void)access;
+  auto* mutexes = static_cast<std::array<std::mutex, CURL_LOCK_DATA_LAST>*>(userptr);
+  mutexes->at(data).lock();
+}
+
+static void curl_share_unlock_cb(CURL* handle, curl_lock_data data, void* userptr) {
+  (void)handle;
+  auto* mutexes = static_cast<std::array<std::mutex, CURL_LOCK_DATA_LAST>*>(userptr);
+  mutexes->at(data).unlock();
+}
+
+void HttpClientWithShare::initCurlShare() {
+  share_ = curl_share_init();
+  if (share_ == nullptr) {
+    throw std::runtime_error("Could not initialize share");
+  }
+
+  curl_share_setopt(share_, CURLSHOPT_SHARE, CURL_LOCK_DATA_SSL_SESSION);
+  curl_share_setopt(share_, CURLSHOPT_SHARE, CURL_LOCK_DATA_DNS);
+  curl_share_setopt(share_, CURLSHOPT_SHARE, CURL_LOCK_DATA_CONNECT);
+  curl_share_setopt(share_, CURLSHOPT_LOCKFUNC, curl_share_lock_cb);
+  curl_share_setopt(share_, CURLSHOPT_UNLOCKFUNC, curl_share_unlock_cb);
+  curl_share_setopt(share_, CURLSHOPT_USERDATA, &curl_share_mutexes);
+  curl_easy_setopt(curl, CURLOPT_SHARE, share_);
```

Comment:
Is this share setting to the original curl handle is required taking into account the handle duplicate is created and share is set for it for each http action/verb?

Reply:
You're right. Since this exact handle is not used directly, there is no need to `curl_easy_setopt(curl, CURLOPT_SHARE, share_);` here. Already tested to make sure, will update the PR in a bit.

Reply:
I suppose no need to move the member field `CURL *curl` to the `protected` area then.